src/models/tools/table/table-content.ts

```diff
@@ -204,101 +235,121 @@ export const TableContentModel = ToolContentModel
+          // The shared model has already been registered by a client, but as the
+          // "owner" of the data, we synchronize it with our local content.
+          if (!self.importedDataSet.isEmpty) {
+            sharedDataSet.dataSet = cloneDeep(self.importedDataSet) as unknown as IDataSet;
+            self.clearImportedDataSet();
+          }
+        }
+        else {
+          if (!sharedDataSet) {
+            // The document doesn't have a shared model yet
+            const dataSet = !self.importedDataSet.isEmpty ? cloneDeep(self.importedDataSet) : DataSet.create();
+            self.clearImportedDataSet();
+            sharedDataSet = SharedDataSet.create({ providerId: self.metadata.id, dataSet });
+          }
+
+          // Add the shared model to both the document and the tile
+          sharedModelManager.addTileSharedModel(self, sharedDataSet);
+        }
+      },
+      {name: "sharedModelSetup", fireImmediately: true}));
+    },
+    updateAfterSharedModelChanges(sharedModel?: SharedModelType) {
+      console.warn("updateAfterSharedModelChanges hasn't been implemented for table content.");
+    },
     setTableName(name: string) {
-      self.dataSet.name = name;
+      self.dataSet.setName(name);
 
-      self.logChange({
-              action: "update",
-              target: "table",
-              props: { name }
-            });
+      self.logChange({ action: "update", target: "table", props: { name } });
     },
-    addAttribute(id: string, name: string, links?: ILinkProperties) {
+    addAttribute(id: string, name: string) {
       self.dataSet.addAttributeWithID({ id, name });
       self.metadata.updateDatasetByExpressions(self.dataSet);
 
-      self.logChange({
-            action: "create",
-            target: "columns",
-            ids: [id],
-            props: { columns: [{ name }] },
-            links
-          });
+      self.logChange({ action: "create", target: "columns", ids: [id], props: { columns: [{ name }] } });
     },
-    setAttributeName(id: string, name: string, links?: ILinkProperties) {
+    setAttributeName(id: string, name: string) {
       const attr = self.dataSet.attrFromID(id);
       if (attr) {
-        attr.name = name;
+        attr.setName(name);
         // if the name of the "x" column is changed, formulas must be updated
         if (self.dataSet.attrIndexFromID(id) === 0) {
           self.metadata.clearRawExpressions(kSerializedXKey);
         }
       }
 
-      self.logChange({
-              action: "update",
-              target: "columns",
-              ids: id,
-              props: { name },
-              links
-            });
+      self.logChange({ action: "update", target: "columns", ids: id, props: { name } });
     },
-    removeAttributes(ids: string[], links?: ILinkProperties) {
+    removeAttributes(ids: string[]) {
       ids.forEach(id => self.dataSet.removeAttribute(id));
       self.metadata.updateDatasetByExpressions(self.dataSet);
```

Comment:
I'm not sure but I suspect all of this use of `metadata` is going to cause problems for the undo and time travel code. I'd like to understand its usage better to see what we can do about it.

Probably this can be left as is in this PR and we can work on that in another PR if necessary.

Reply:
The metadata came into existence at a time when document changes frequently resulted in `DocumentContent` being disposed and recreated from scratch and certain pending changes would be lost. It evolved to also be used to support synchronization between multiple document instances. The former use case is probably no longer relevant since we switched to using `applySnapshot()` rather than replacing the content entirely. The metadata may end up being useful again if we end up splitting read-only and editable document instances due to history scrubbing, because it would allow some information to be shared between document instances without having to be synchronized through firebase. Definitely some issues to work through here.